Sources/SWBBuildService/BuildDescriptionMessages.swift

```diff
@@ -0,0 +1,218 @@
+        let buildRequestContext = BuildRequestContext(workspaceContext: workspaceContext)
+        let clientDelegate = ClientExchangeDelegate(request: self, session: session)
+        let operation = IndexingOperation(workspace: workspaceContext.workspace)
+        let buildDescription = try await session.buildDescriptionManager.getNewOrCachedBuildDescription(
+            .cachedOnly(
+                message.buildDescriptionID,
+                request: buildRequest,
+                buildRequestContext: buildRequestContext,
+                workspaceContext: workspaceContext
+            ), clientDelegate: clientDelegate, constructionDelegate: operation
+        )?.buildDescription
+
+        guard let buildDescription else {
+            throw BuildDescriptionDoesNotExistError()
+        }
+        return (buildRequest, buildDescription)
+    }
+}
+
+// MARK: - Message handlers
+
+struct BuildDescriptionConfiguredTargetsMsg: MessageHandler {
+    /// Compute the toolchains that can handle all Swift and clang compilation tasks in the given target.
+    private func toolchainIDs(in configuredTarget: ConfiguredTarget, of buildDescription: BuildDescription) -> [String]? {
+        var toolchains: [String]?
+
+        for task in buildDescription.taskStore.tasksForTarget(configuredTarget) {
+            let targetToolchains: [String]? =
+                switch task.payload {
+                case let payload as SwiftTaskPayload: payload.indexingPayload.toolchains
+                case let payload as ClangTaskPayload: payload.indexingPayload?.toolchains
+                default: nil
+                }
+            guard let targetToolchains else {
+                continue
+            }
+            if let unwrappedToolchains = toolchains {
+                toolchains = unwrappedToolchains.filter { targetToolchains.contains($0) }
+            } else {
+                toolchains = targetToolchains
+            }
+        }
+
+        return toolchains
+    }
+
+    func handle(request: Request, message: BuildDescriptionConfiguredTargetsRequest) async throws -> BuildDescriptionConfiguredTargetsResponse {
+        let buildDescription = try await request.buildDescription(for: message)
+
+        let dependencyRelationships = Dictionary(
+            buildDescription.targetDependencies.map { (ConfiguredTarget.GUID(id: $0.target.guid), [$0]) },
+            uniquingKeysWith: { $0 + $1 }
+        )
+
+        let session = try request.session(for: message)
+
+        let targetInfos = buildDescription.allConfiguredTargets.map { configuredTarget in
+            let toolchain: Path?
+            if let toolchainID = toolchainIDs(in: configuredTarget, of: buildDescription)?.first {
+                toolchain = session.core.toolchainRegistry.lookup(toolchainID)?.path
+                if toolchain == nil {
+                    log("Unable to find path for toolchain with identifier \(toolchainID)", isError: true)
```

Comment:
@ahoppen We don't conventionally have this level of logging in the product; it seems like debug printing that should've been removed.